### PR TITLE
docs/contributing: Add Versioned Resources section to Provider Design page

### PR DESCRIPTION
**Pull request by bflad**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Reference: https://github.com/hashicorp/terraform-provider-aws/pull/17034#issuecomment-766193857

Output from acceptance testing: N/A (documentation)

### CODE REVIEW

**Review by ewbankkit — APPROVED** [review]
LGTM.
`aws_s3_bucket_object` violates most of these guidelines 😭.
The resource represents an object without any version (or version is `null`) if the object is in a non-versioned bucket and the latest version if the object is in a versioned bucket, but then deletes all versions of the object when the resource is destroyed.

File: docs/contributing/provider-design.md [hunk 1]
```diff
@@ -118,6 +119,20 @@ In this situation, provider developers should create a separate resource represe
 
 For a related consideration, see the [Managing Resource Running State section](#managing-resource-running-state).
 
+### Versioned Resources
+
+Some AWS components are versioned, where the API supports multiple configurations of a component. Examples of this include:
```
Comment (YakDriver) [inline docs/contributing/provider-design.md:124]:
```suggestion
AWS supports having multiple versions of some components. Examples of this include:
```

File: docs/contributing/provider-design.md [hunk 2]
```diff
@@ -118,6 +119,20 @@ In this situation, provider developers should create a separate resource represe
...
+
+* ECS Task Definitions
+* Lambda Functions
+* Secrets Manager Secrets
+
+In this situation, provider developers should prefer creating a separate resource to represent a single version of the component. General guidelines for deciding when to create a separate resource or not:
```
Comment (YakDriver) [inline docs/contributing/provider-design.md:130]:
```suggestion
In general, provider developers should create a separate resource to represent a single version. For example, the provider has both the `aws_secretsmanager_secret` and `aws_secretsmanager_secret_version` resources. However, in some cases, developers should handle versioning in the main resource.

In deciding when to create a separate resource, follow these guidelines:
```

File: docs/contributing/provider-design.md [hunk 3]
```diff
@@ -118,6 +119,20 @@ In this situation, provider developers should create a separate resource represe
...
+
+* If resource creation must create a version of that resource, then typically the Terraform resource should be self-contained with versioning. Semantically it would be confusing to have one resource to manage something at the beginning then a separate resource for later updates.
```
Comment (YakDriver) [inline docs/contributing/provider-design.md:132]:
```suggestion
* If AWS necessarily creates a version when you make a new AWS component, include version handling in the same Terraform resource. Creating an AWS component with one Terraform resource and later using a different resource for updates is confusing.
```

File: docs/contributing/provider-design.md [hunk 4]
```diff
@@ -118,6 +119,20 @@ In this situation, provider developers should create a separate resource represe
...
+* If version deletion is possible or necessary, then typically there should be separate resources.
```
Comment (YakDriver) [inline docs/contributing/provider-design.md:133]:
```suggestion
* If the AWS service API allows deleting versions and practitioners will want to delete versions, provider developers should implement a separate version resource.
```

File: docs/contributing/provider-design.md [hunk 5]
```diff
@@ -118,6 +119,20 @@ In this situation, provider developers should create a separate resource represe
...
+* If the API only supports publishing new versions, either method is acceptable, however most current implementations are self-contained. Terraform's current configuration language does not natively support triggering resource updates or recreation across resources without a state value change. This can make the implementation more difficult for practitioners without special resource and configuration workarounds, such as a `triggers` attribute. If this changes in the future, then this guidance may be updated towards separate resources, following the [Task Execution and Waiter Resources](#task-execution-and-waiter-resources) guidance.
```
Comment (YakDriver) [inline docs/contributing/provider-design.md:134]:
```suggestion
* Where the API only supports publishing new versions, developers can use either approach.
* Note that, in contrast to this guidance, in the past, contributors often implemented version handling within the primary resource.
* Keep in mind that Terraform's HCL does not natively support triggering resource updates or recreation across resources without a state value change. The lack complicates configurations for practitioners and requires them to use workarounds, such as a `triggers` attribute. When this changes in the future, we will update this guide to recommend implementing separate version resources, following the [Task Execution and Waiter Resources](#task-execution-and-waiter-resources) guidance.
```

**Review by YakDriver — APPROVED** [review]
Great information!

**Pull request merged**